src/Microsoft.AspNetCore.Blazor/Rendering/Renderer.cs

```diff
@@ -113,7 +118,28 @@ internal void InstantiateChildComponentOnFrame(ref RenderTreeFrame frame)
         internal void AssignEventHandlerId(ref RenderTreeFrame frame)
         {
             var id = ++_lastEventHandlerId;
-            _eventHandlersById.Add(id, (UIEventHandler)frame.AttributeValue);
+
+            // The attribute value might be a more specialized type like UIKeyboardEventHandler.
+            //
+            // In order to dispatch the event, we need a UIEventHandler, so we're going weakly
+            // typed here. The user will get a cast exception if they map the wrong type of
+            // delegate to the event.
+            if (frame.AttributeValue is UIEventHandler wrapper)
+            {
+                _eventHandlersById.Add(id, wrapper);
+            }
+            else if (frame.AttributeValue is MulticastDelegate @delegate)
+            {
+                // IMPORTANT: we're creating an additional delegate when necessary. This is
+                // going to get cached in _eventHandlersById, but the render tree diff
+                // will operate on 'AttributeValue' which means that we'll only create a new
+                // wrapper delegate when the underlying delegate changes.
+                //
+                // TLDR: If the component uses a method group or a non-capturing lambda
+                // we don't allocate much.
+               _eventHandlersById.Add(id, (UIEventArgs e) => @delegate.DynamicInvoke(e));
+            }
```

Comment:
This was the trick. We only generate a 'wrapper' after we've processed the diff and decided to register the delegate for events. This allows the diff phase to leverage the delegate equality rules

Reply:
Makes sense. Sounds like a good way to get the perf benefits.